miasm/expression/simplifications_common.py

```diff
@@ -1753,19 +1753,39 @@ def simp_compose_and_mask(_, expr):
     if not arg2.is_int():
         return expr
     int2 = int(arg2)
-    if (int2 + 1) & int2 != 0:
-        return expr
-    mask_size = int2.bit_length() + 7 // 8
+    mask_size = (int2.bit_length() + 7) // 8 * 8
+    if int2 == int(arg1.mask):
+        return arg1
     out = []
+    mask_needed = False
     for offset, arg in arg1.iter_args():
         if offset == mask_size:
-            return ExprCompose(*out).zeroExtend(expr.size)
-        elif mask_size > offset and mask_size < offset+arg.size and arg.is_int():
-            out.append(ExprSlice(arg, 0, mask_size-offset))
-            return ExprCompose(*out).zeroExtend(expr.size)
+            break
```

Comment:
Ok, maybe here, instead of a break, juste append a `ExprInt(0, arg.size)`
This way, you generate generate the ExprCompose with the final elements, and don't need to have special cases at the end of the test.
More over, if I don't mess up, you may result with
```
{XXXX, 0, 8, ExprInt(0, 8), 8, 16, ExprInt(0, 16), 16, 32} 
```
But this will be simplified by another rule into:
```
{XXX, 0, 8, ExprInt(0, 24), 8, 32}
```

Reply:
i think this is true